logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -0,0 +1,385 @@
+        inputs = setupInputs();
+        filters = setupFilters();
+        outputs = setupOutputs();
+        shutdownFlushes = Collections.unmodifiableList(
+            filters.values().stream().filter(RubyIntegration.Filter::hasFlush)
+                .collect(Collectors.toList())
+        );
+        periodicFlushes = Collections.unmodifiableList(
+            shutdownFlushes.stream().filter(RubyIntegration.Filter::periodicFlush)
+                .collect(Collectors.toList())
+        );
+    }
+
+    public Collection<RubyIntegration.Filter> shutdownFlushers() {
+        return shutdownFlushes;
+    }
+
+    public Collection<RubyIntegration.Filter> periodicFlushers() {
+        return periodicFlushes;
+    }
+
+    public Collection<RubyIntegration.Output> outputs() {
+        return Collections.unmodifiableCollection(outputs.values());
+    }
+
+    public Collection<RubyIntegration.Filter> filters() {
+        return Collections.unmodifiableCollection(filters.values());
+    }
+
+    public Collection<IRubyObject> inputs() {
+        return inputs;
+    }
+
+    public RubyIntegration.Plugin registerPlugin(final RubyIntegration.Plugin plugin) {
+        plugin.register();
+        return plugin;
+    }
+
+    /**
+     * This method contains the actual compilation of the {@link Dataset} representing the
+     * underlying pipeline from the Queue to the outputs.
+     * @return Compiled {@link Dataset} representation of the underlying {@link PipelineIR} topology
+     */
+    public Dataset buildExecution() {
+        return new CompiledPipeline.CompiledExecution().toDataset();
+    }
+
+    /**
+     * Sets up all Ruby outputs learnt from {@link PipelineIR}.
+     */
+    private Map<String, RubyIntegration.Output> setupOutputs() {
+        final Collection<PluginVertex> outs = pipelineIR.getOutputPluginVertices();
+        final Map<String, RubyIntegration.Output> res = new HashMap<>(outs.size());
+        outs.forEach(v -> {
+            final PluginDefinition def = v.getPluginDefinition();
+            final SourceWithMetadata source = v.getSourceWithMetadata();
+            res.put(v.getId(), pipeline.buildOutput(
+                RubyUtil.RUBY.newString(def.getName()), RubyUtil.RUBY.newFixnum(source.getLine()),
+                RubyUtil.RUBY.newFixnum(source.getColumn()), convertArgs(def)
+            ));
+        });
+        return res;
+    }
+
+    /**
+     * Sets up all Ruby filters learnt from {@link PipelineIR}.
+     */
+    private Map<String, RubyIntegration.Filter> setupFilters() {
+        final Collection<PluginVertex> filterPlugins = pipelineIR.getFilterPluginVertices();
+        final Map<String, RubyIntegration.Filter> res =
+            new HashMap<>(filterPlugins.size(), 1.0F);
+        for (final PluginVertex plugin : filterPlugins) {
+            final String ident = plugin.getId();
+            if (!res.containsKey(ident)) {
+                res.put(ident, buildFilter(plugin));
+            }
```

Comment:
**ids** should be unique per plugin, if it's not the case we shouldn't silently skip them. Also inputs/output should have the same logic? Maybe we should have a more complex type than a hashmap to act as a plugin registry which would ensure ids are unique?

Reply:
Also we have similar logic for setupInput/setupOutput, We could probably extract the pattern?

https://github.com/elastic/logstash/pull/7950/files#diff-484ab4968ddf73de6d56cb55c3b01ed0R128

Reply:
@ph actually that check is redundant, let me remove it :)

Reply:
@ph the current approach is basically a result of me not wanting to have the Ruby constant strings for "filter", "output" and so on on the Java side. I fear that the way I call `pipeline.buildFilter` and the other can't really be dried up without a dynamic method call, this is where Java is just Java, it annoyed me as well :)